src/librustc_incremental/calculate_svh/svh_visitor.rs

```diff
@@ -17,24 +17,99 @@ use self::SawExprComponent::*;
 use syntax::parse::token;
-use syntax_pos::Span;
+use syntax::codemap::CodeMap;
+use syntax_pos::{Span, NO_EXPANSION, COMMAND_LINE_EXPN, BytePos, FileMap};
 use rustc::hir;
 use rustc::hir::*;
 use rustc::hir::def::{Def, PathResolution};
 use rustc::hir::def_id::DefId;
 use rustc::hir::intravisit as visit;
 use rustc::ty::TyCtxt;
-
+use std::rc::Rc;
 use std::hash::{Hash, SipHasher};
 
 use super::def_path_hash::DefPathHashes;
 
+const IGNORED_ATTRIBUTES: &'static [&'static str] = &["cfg",
+                                                      "rustc_clean",
+                                                      "rustc_dirty"];
+
 pub struct StrictVersionHashVisitor<'a, 'hash: 'a, 'tcx: 'hash> {
     pub tcx: TyCtxt<'hash, 'tcx, 'tcx>,
     pub st: &'a mut SipHasher,
     // collect a deterministic hash of def-ids that we have seen
     def_path_hashes: &'a mut DefPathHashes<'hash, 'tcx>,
     hash_spans: bool,
+    codemap: CachedCodemapView<'tcx>,
+}
+
+struct CachedCodemapView<'tcx> {
+    codemap: &'tcx CodeMap,
+    // Format: (line number, line-start, line_end, file)
+    line_cache: [(usize, BytePos, BytePos, Rc<FileMap>); 4],
+    eviction_index: usize,
+}
+
+impl<'tcx> CachedCodemapView<'tcx> {
+    fn new<'a>(tcx: TyCtxt<'a, 'tcx, 'tcx>) -> CachedCodemapView<'tcx> {
+        let codemap = tcx.sess.codemap();
+        let first_file = codemap.files.borrow()[0].clone();
+
+        CachedCodemapView {
+            codemap: codemap,
+            line_cache: [(0, BytePos(0), BytePos(0), first_file.clone()),
+                         (0, BytePos(0), BytePos(0), first_file.clone()),
+                         (0, BytePos(0), BytePos(0), first_file.clone()),
+                         (0, BytePos(0), BytePos(0), first_file.clone())],
+            eviction_index: 0,
+        }
+    }
+
+    fn byte_pos_to_line_and_col(&mut self,
+                                pos: BytePos)
+                                -> (Rc<FileMap>, usize, BytePos) {
+        // Check if the position is in one of the cached lines
+        for &(line, start, end, ref file) in self.line_cache.iter() {
+            if pos >= start && pos < end {
+                return (file.clone(), line, pos - start);
+            }
+        }
+
+        // Check whether we have a cached line in the correct file, so we can
+        // overwrite it without having to look up the file again.
+        for &mut (ref mut line,
+                  ref mut start,
+                  ref mut end,
+                  ref file) in self.line_cache.iter_mut() {
+            if pos >= file.start_pos && pos < file.end_pos {
+                let line_index = file.lookup_line(pos).unwrap();
+                let (line_start, line_end) = file.line_bounds(line_index);
+
+                // Update the cache entry in place
+                *line = line_index + 1;
+                *start = line_start;
+                *end = line_end;
+
+                return (file.clone(), line_index + 1, pos - line_start);
```

Comment:
I think this strategy will lead to you re-using the same cache line over and over (e.g., if you have just 1 file, you'll load some data into slot 0, then update slot 0 to be the next line and the next...). I'm not sure this is optimal. Probably in many use cases you never use a line once we've moved past it, but macro expansion changes that calculus some also.